tensorflow/c/eager/dlpack.cc

```diff
@@ -0,0 +1,332 @@
+        case 16:
+          tf_dtype = TF_DataType::TF_UINT16;
+          break;
+        case 32:
+          tf_dtype = TF_DataType::TF_UINT32;
+          break;
+        case 64:
+          tf_dtype = TF_DataType::TF_UINT64;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported UInt bits: ", dtype.bits);
+      }
+      break;
+    case DLDataTypeCode::kDLInt:
+      switch (dtype.bits) {
+        case 8:
+          tf_dtype = TF_DataType::TF_INT8;
+          break;
+        case 16:
+          tf_dtype = TF_DataType::TF_INT16;
+          break;
+        case 32:
+          tf_dtype = TF_DataType::TF_INT32;
+          break;
+        case 64:
+          tf_dtype = TF_DataType::TF_INT64;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported Int bits: ", dtype.bits);
+      }
+      break;
+    case DLDataTypeCode::kDLFloat:
+      switch (dtype.bits) {
+        case 16:
+          tf_dtype = TF_DataType::TF_HALF;
+          break;
+        case 32:
+          tf_dtype = TF_DataType::TF_FLOAT;
+          break;
+        case 64:
+          tf_dtype = TF_DataType::TF_DOUBLE;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported Float bits: ", dtype.bits);
+      }
+      break;
+    case DLDataTypeCode::kDLBfloat:
+      switch (dtype.bits) {
+        case 16:
+          tf_dtype = TF_DataType::TF_BFLOAT16;
+          break;
+        default:
+          status->status = tensorflow::errors::InvalidArgument(
+              "Unsupported BFloat bits: ", dtype.bits);
+      }
+      break;
+    default:
+      status->status = tensorflow::errors::InvalidArgument(
+          "Unsupported Type Codes: ", dtype.code);
+  }
+
+  return tf_dtype;
+}
+
+// Wrapper function to match the function signature
+// TFE_NewTensorHandleFromDeviceMemory, calling the deleter of the
+// DLManagedTensor
+void DeallocatorWrapperFunc(void* data, size_t len, void* dlmt_vptr) {
+  DLManagedTensor* dlmt = static_cast<DLManagedTensor*>(dlmt_vptr);
+  dlmt->deleter(const_cast<DLManagedTensor*>(dlmt));
+}
+
+// Check whether the stride array matches the layout of compact, row-majored
+// data
+bool IsValidStrideCompactRowMajorData(int64_t* shape_arr, int64_t* stride_arr,
+                                      int ndim) {
+  if (ndim >= 1 && stride_arr[ndim - 1] != 1) {
+    return false;
+  }
+  for (int i = ndim - 2; i >= 0; --i) {
+    if (stride_arr[i] != shape_arr[i + 1] * stride_arr[i + 1]) {
+      return false;
+    };
+  }
+  return true;
+}
+}  // namespace
+
+void TFE_CallDLManagedTensorDeleter(void* dlm_ptr) {
+  DLManagedTensor* dlMTensor = static_cast<DLManagedTensor*>(dlm_ptr);
+  if (dlMTensor->deleter != nullptr) {
+    dlMTensor->deleter(dlMTensor);
+  }
+}
+
+void* TFE_HandleToDLPack(TFE_TensorHandle* h, TF_Status* status) {
```

Comment:
Sorry for noticing this so late, but in the design review we discussed that we needed to do a host/device sync when converting to/from DLPack tensors, via a hypothetical `TFE_SynchronizeDevice`.  Is that not needed?

Reply:
I think the sync is done in the `TFE_TensorHandleDevicePointer`
https://github.com/VoVAllen/tensorflow/blob/dlpack/tensorflow/c/eager/c_api.cc#L1145-L1180

https://github.com/VoVAllen/tensorflow/blob/dlpack/tensorflow/c/eager/dlpack.cc#L269

Reply:
I see.  How about in `TFE_HandleFromDLPack`?  Or is that not necessary because dlpack will ensure that the memory is available to read on any stream?

Reply:
I think based on our previous discussion, the framework creating dlpack capsule should ensure the availability of the memory. Is this the case @jermainewang ?

Reply:
Ok, thanks.